### ISSUE TITLE
Unexpected error for argument with type partial generic and intersection

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#27473]
**TypeScript Version:**  3.0.3

**Search Terms:** generic intersection partial cloneElement React

**Code**

```ts
type Attributes = {
    key: string;
};

type ReactElement<P> = { props: P };

declare const cloneElement: <P>(
    element: ReactElement<P>,
    props: Attributes & Partial<P>,
) => ReactElement<P>;

const addKey = (key: string) => <P>(el: ReactElement<P>) =>
    cloneElement(
        el,
        /*
        Argument of type '{ key: string; }' is not assignable to parameter of type 'Attributes & Partial<P>'.
            Type '{ key: string; }' is not assignable to type 'Partial<P>'.
        */
        { key },
    );
```

(This is necessary to model React's `cloneElement` function.)

**Comment by dhruvrajvanshi** [comment]
I'm not sure what you're trying to do here.

I'm assuming you're trying to clone and element and replace a prop named `key` with the passed argument.
It's not safe because `P` may already contain the key `key` which may have a different type than `string`.
Your component that's being cloned expects its props to be of type P and if you assign another string to its props, it may break.

**Comment by OliverJAsh** [comment]
> I'm assuming you're trying to clone and element and replace a prop named `key`

`key` is a special, built-in prop for React components.

In the same way it's possible to pass `key` as a prop to a component (where it doesn't define a custom `key` prop), I would like to clone a React element to override it's `key` prop (which all components/elements have).

``` tsx
const MyComponent: SFC<{ name: string; }> = ({ name }) => <div>{name}</div>

<MyComponent name="bob" key="foo />
```

The React side of things doesn't really matter here. The issue is that I can't assign `{ key: string }` to `Partial<P> & { key: string }`, when I believe I should be able to.

**Comment by dhruvrajvanshi** [comment]
It is correct behaviour with regards to type safety though. The type of `key` in `{  key: string } & Partial<P>` is `P['key'] & string` and it makes sense that `string` isnt assignable  to that (maybe P['key'] is number for some P). Maybe try adding `extends {  key: string }` constraint to P in cloneElement?

**Comment by jack-williams** [comment]
The constraint on `P` will not work because `P` could be `never`. Without lower bounding I'm not sure there is a safe way to write this.

**Comment by RyanCavanaugh** [comment]
Other commentors are correct; you have knowledge (`key` won't appear in `props`) that TypeScript isn't able to reason about. A type assertion is appropriate to use here.
